### ISSUE TITLE
CHMFSampler silently mis-normalizes when the conditioning field isn't pre-smoothed to sigma^2(M_env)

### ISSUE THREAD

**Issue by sambit-giri** [cosmic-reionization/BEoRN#54]
## Summary

`CHMFSampler.sample()` / `expected_counts()`, in the default per-cell conditioning mode (`R_env=None`), silently assumes the `delta_field` it is handed already has variance `sigma^2(M_env, z)`, where `M_env = rho_m * cell_size**3` is derived internally from `parameters`. There is no check and no smoothing applied in this branch (`CHMFSampler._environment`, `chmf.py`):

```python
def _environment(self, delta_field, R_env):
    params = self.parameters
    cell_size = params.Lbox_hunits / params.simulation.Ncell
    if R_env is None:
        return delta_field, self.chmf.rho_m * cell_size ** 3
    ...
```

If the caller passes the *raw* linear field (e.g. `LPTBase.get_linear_density(z)` with `R_tophat=None`, or any field not explicitly top-hat-smoothed to the cell-equivalent radius), the field's actual variance does not match the analytic `sigma^2(M_env, z)` used internally — but nothing signals this. The conditional HMF is systematically biased as a result: EPS self-consistency (documented in `LPTBase.get_linear_density`'s own docstring — "EPS self-consistency requires the conditioning field to be Gaussian with variance sigma^2(M_env)") is silently violated.

## Reproduction / measured impact

Using the differentiable branch, `Ncell=256`, `Lbox=200` physical Mpc (`h0=0.68`), fiducial Planck-like cosmology:

- cell size = 0.531 Mpc/h, `M_env` = 1.897e10 Msun, analytic `sigma^2(M_env, z=11)` = 0.1514
- raw (unsmoothed) `get_linear_density(z=11)` field variance = 0.1835 (**21% too high**)
- the same field top-hat-smoothed at the correct equivalent radius (`R_eq = cell_size*(3/(4*pi))**(1/3) = 0.330` Mpc/h) has variance 0.1425 (within ~6% of the analytic value — residual is finite-box/grid discreteness, not the smoothing mismatch)

Sampling a discrete catalog with `CHMFSampler(hmf_model='ST').sample(...)` from the *unsmoothed* field and volume-averaging it (the EPS self-consistency check: this should reproduce the unconditional ST curve) gives a systematic excess over the analytic ST `dn/dlnM` of **19.7% (z=15), 11.8% (z=11), 5.6% (z=7)** (mean `|catalog/ST - 1|` across the mass range) — entirely explained by the variance mismatch above, not Poisson/shot noise.

Separately (not a bug, but worth documenting alongside this): per-cell conditioning also hard-caps the maximum sampleable halo mass at `0.999*M_env` (`_mass_bins`) — for the box above that's ~1.9e10 Msun, well below what a wide `M_bins` range might suggest is being sampled.

## Suggested fix

In the `R_env is None` branch of `_environment` (and equivalently in `expected_counts`/`sample`), apply the correct top-hat smoothing (`R_eq = cell_size*(3/(4*pi))**(1/3)`) to `delta_field` internally, rather than assuming the caller has already done it. This removes the silent "caller must pre-smooth to exactly the right scale" contract and closes the gap for any future caller (not just this one case). Needs a decision on backward compatibility: this changes behavior for any existing caller currently passing an already-correctly-smoothed field (double smoothing) — may need a flag (e.g. `assume_presmoothed=False` default) or should simply become the only supported path with a major/minor version bump, documented clearly in `sample()`'s docstring.

A cheap complementary safety net regardless of the fix chosen: a debug-level warning if `delta_field.var()` differs from `sigma^2(M_env, z)` by more than some tolerance (e.g. 10%), to catch this class of mistake even when a caller does something unexpected.

**Comment by sambit-giri** [comment]
Fix implemented (currently on the `differentiable` branch, working tree — not yet committed/PR'd).

**Root cause, precisely:** `CHMFSampler._environment()` never smoothed `delta_field` at all in the default `R_env=None` path (returned it completely unmodified), and in the explicit `R_env > cell_size` path it re-smoothed an already-correctly-smoothed field (`LPTHaloLoader` pre-smooths itself) a second time with an inconsistent **Gaussian** window — a real, live double-smoothing bug whenever `halo_sim.R_env` was set above the cell size.

**Fix:** no backward-compatibility flag — `CHMFSampler` now always top-hat-smooths its own conditioning field internally to the correct radius (`R_eq` by default, or `R_env` when set), via a new `_tophat_smooth` helper built on the existing `TopHatWindow` class (consistent with the top-hat window `sigma2(M,z)` already assumes everywhere else). The old Gaussian `_smooth_field` is deleted. `LPTHaloLoader` no longer pre-smooths itself — it just hands over the raw linear density field, since smoothing is now 100% `CHMFSampler`'s responsibility.

**Deliberate default-behavior change:** anyone using `halo_sim.R_env` above the cell size will get numerically different (corrected) catalogs — the old behavior was simply wrong, so nothing is preserved for backward compatibility there.

**Test coverage added** (`tests/test_chmf.py`): variance-matching checks for both the default and explicit-`R_env` conditioning paths, plus an EPS self-consistency test (volume-averaged conditional HMF vs. analytic Press-Schechter on a raw, unsmoothed field) — the actual bug-closure test, which fails on the pre-fix code and passes after.

Full local suite: 398 passed, 3 skipped, 14 deselected (known MPI/GPU-hardware-only exclusions unrelated to this change).

**Comment by sambit-giri** [comment]
Pushed in 4945fe6 on the `differentiable` branch (same commit as #56 — the fix builds directly on that Parameters restructuring, so they landed together). Leaving this open for now, more updates still planned.

**Comment by sambit-giri** [comment]
Confirmed ready: the always-smooth fix (commit 4945fe6) is intact in the current code, and its three regression tests (variance-matching for both the default and explicit-R_env conditioning paths, plus the EPS self-consistency volume-average check) pass in the full local suite (408 passed).

Note: this fix lives on the `differentiable` branch, not yet merged into `main` -- closing now per your call, but it isn't live for anyone on `main` until that merge happens.

**Comment by sambit-giri** [comment]
Follow-up note (unrelated to this issue's own bug, logged here since it touches the same `chmf.py`/CHMF module): pushed `commit 05ad9df`, porting `CHMF.hmf_st_movingbarrier` (the Davies, Mesinger & Murray 2025 moving-barrier conditional HMF, eqs. 2-4) to the differentiable `conditional_dndlnm_diff`/`halo_field_diff` path via a new `chmf_recipe` argument -- previously only available on the numpy reference class. Verified against the numpy version (rtol=1e-3) and `jax.grad` matches central finite differences.

Context: this is prep work for reviewing `CHMFSampler` against that same paper's DexM+CHMF hybrid halo-generation approach (massive halos identified deterministically from the density field, low-mass halos sampled stochastically from the CHMF) -- a design discussion still in progress, not yet a concrete plan.
